### PR TITLE
Refactor in favor to use `primitiveOrStruct`.

### PR DESCRIPTION
**Pull request by ribeiromiranda**
Refactor in favor to use `primitiveOrStruct`.

### CODE REVIEW

**Comment by ribeiromiranda** [comment]
Not found in the specification, what is the motivation for function `unmarshalFile`?

When a parameter can be a `string` or `object`, must always check if a string is a file path and try loading?

**Comment by ricardozanini** [comment]
@ribeiromiranda sometimes a reference can be an external file instead of a plain JSON object. In these cases, we need to load the file, process it and unmarshal it into a struct.

**Comment by spolti** [comment]
@ribeiromiranda hi, can you please sign your commit and rebase the PR?
Thanks.

File: model/util.go
Comment (spolti) [inline model/util.go:124]: what is the purpose of it?

Comment (ribeiromiranda) [reply model/util.go:124]: To write unit tests with httptest. Example in the `fileunit_test.go` `Test_getBytesFromFile`.

Comment (spolti) [reply model/util.go:124]: yes, but why is it needed?

Comment (ribeiromiranda) [reply model/util.go:124]: Following the specification rule. "Referencing a reusable JSON/Yaml error definition file" an example the [defining errors](https://github.com/serverlessworkflow/specification/blob/0.8.x/specification.md#defining-errors
) 

What schemes can the URI be: `file://`, `https://`, `ftp://`?

Comment (spolti) [reply model/util.go:124]: I see, but I think it is up to the engine to implement it, no?
@ricardozanini wdyt?

Comment (ricardozanini) [reply model/util.go:124]: No, this is our duty since the external file contains the definition for the object. For example:

```json
{
    "functions": "myfunctions.json"
}
```

The structure should be `Workflow().Functions()[0].Name`. Not `Workflow().Functions()` ~ a string

@ribeiromiranda the scheme should be determined by a URI parser. By default, it's ` file://` if the scheme is empty. If remote like HTTP(S), we do fetch.

**Comment by spolti** [comment]
Hi  @ribeiromiranda, can you please rebase?

File: parser/testdata/workflows/applicationrequest-issue103.json
Comment (spolti) [inline parser/testdata/workflows/applicationrequest-issue103.json:8]: can't we keep without protocol?
maybe add just into one to make sure the unmarshal will work? wdyt?

**Comment by spolti** [comment]
@lsytj0413 hi, if you have time, do you mind reviewing this PR?
Thanks.

File: model/action.go
Comment (lsytj0413) [inline model/action.go:110]: All public Method with exported Type should have an comment

File: model/action_data_filter_test.go
Comment (lsytj0413) [inline model/action_data_filter_test.go:64]: an string ->  a string

File: model/util.go
Comment (lsytj0413) [inline model/util.go:54]: replace \"%s\" with %q?

File: model/util.go
Comment (lsytj0413) [inline model/util.go:60]: replace 0 with reflect.Invalid？

File: model/util.go
Comment (lsytj0413) [inline model/util.go:63]: replace 0 with reflect.Invalid？

File: model/util.go
Comment (lsytj0413) [inline model/util.go:83]: This will be more complicated when adding a new type, and synchronous modifications are also needed here.

File: model/util.go
Comment (lsytj0413) [inline model/util.go:173]: This function's implement is same as unmarshalObjectOrFile，can we just keep one of them？

File: parser/testdata/applicationrequestfunctions.json
Comment (lsytj0413) [inline parser/testdata/applicationrequestfunctions.json:2]: why we remove the outer `function` properties？in spec 0.8，the reference resource should be an object (not an array), this is describe in:

- https://github.com/serverlessworkflow/specification/blob/0.8.x/specification.md#workflow-definition-structure

File: model/event_state.go
Comment (spolti) [inline model/event_state.go:57]: comment?

File: model/event_state.go
Comment (spolti) [inline model/event_state.go:84]: same

File: model/event_test.go
Comment (spolti) [inline model/event_test.go:52]:
```suggestion
			err:    `eventRef.invoke must be sync or async`,
```
can this message be updated to the example above?

File: model/event_test.go
Comment (spolti) [inline model/event_test.go:102]:
```suggestion
			err:    `event.dataOnly must be bool`,
```

File: model/foreach_state_test.go
Comment (spolti) [inline model/foreach_state_test.go:52]:
```suggestion
			err:    `forEachState.mode must be sequential or parallel`,
```

**Review by spolti — CHANGES_REQUESTED** [review]
Nice refactoring :D
Few comments

File: .github/workflows/Go-SDK-Check-k8s-integration.yaml
Comment (ribeiromiranda) [reply .github/workflows/Go-SDK-Check-k8s-integration.yaml:26]: Inside function `unmarshalObject` define the `type Unmarshal *U` and works with go >= 1.20

Comment (spolti) [reply .github/workflows/Go-SDK-Check-k8s-integration.yaml:26]: Does it works or *only* works with 1.20+?

Comment (ribeiromiranda) [reply .github/workflows/Go-SDK-Check-k8s-integration.yaml:26]: Exactly.

Version 1.19 the error occurs:
```
type declarations inside generic functions are not currently supported
```

Comment (spolti) [reply .github/workflows/Go-SDK-Check-k8s-integration.yaml:26]: I see, wee need to be cautions on situations like these, because there could be users that would be using older go lang versions and can't update it.

@ricardozanini @lsytj0413 wdyt?

Comment (lsytj0413) [reply .github/workflows/Go-SDK-Check-k8s-integration.yaml:26]: Agree with @spolti ，go1.19 is still under active supported，so we should promise this sdk will working with go1.19

File: model/util.go
Comment (ribeiromiranda) [reply model/util.go]: The last commit `029d335` I add an interface to deal with validations and error handling.
```
type Kinds interface {
	AllKinds() []string
	String() string
}
```
That way the validation and error handling can get the same values in the method `AllKinds`.

File: model/util.go
Comment (ribeiromiranda) [reply model/util.go:166]: Keeping the function's name I think is important, but I did a refactoring, to avoid repetition.

File: parser/testdata/applicationrequestfunctions.json
Comment (ribeiromiranda) [reply parser/testdata/applicationrequestfunctions.json:2]: The spec have this definition

> The functions property can be either an in-line [function](https://github.com/serverlessworkflow/specification/blob/0.8.x/specification.md#Function-Definition) definition array, or an URI reference to a resource containing an array of [functions](https://github.com/serverlessworkflow/specification/blob/0.8.x/specification.md#Function-Definition) definition. Referenced resource can be used by multiple workflow definitions.

The [sdk-net](https://github.com/serverlessworkflow/sdk-net/blob/main/src/ServerlessWorkflow.Sdk.UnitTests/Resources/functions/petstore.json) also follow that definition

Comment (spolti) [reply parser/testdata/applicationrequestfunctions.json:2]: If we are going to update it, I suggest drop it from this PR and open a new issue to address/discuss it.

File: model/util.go
Comment (ribeiromiranda) [reply model/util.go:88]: It will always be a pointer because of the previous line.
```
val := reflect.New(unmarshalTypeErr.Type)
```

File: parser/testdata/applicationrequestfunctions.json
Comment (ribeiromiranda) [reply parser/testdata/applicationrequestfunctions.json]: Add the ability to unmarshal both cases object or array. What do you think?

Also, opened the issue #168

File: .github/workflows/Go-SDK-Check-k8s-integration.yaml
Comment (ribeiromiranda) [reply .github/workflows/Go-SDK-Check-k8s-integration.yaml]: Solved the support to go 1.19.

File: parser/testdata/applicationrequestfunctions.json
Comment (ribeiromiranda) [reply parser/testdata/applicationrequestfunctions.json]: Back the external resource as object.

File: parser/testdata/eventbasedgreetingevents.json
Comment (spolti) [inline parser/testdata/eventbasedgreetingevents.json:8]:
```suggestion
  ]
```

File: parser/testdata/errors.json
Comment (spolti) [inline parser/testdata/errors.json:2]: any specific reason to change this?

File: parser/testdata/functiondefs.json
Comment (spolti) [inline parser/testdata/functiondefs.json:15]:
```suggestion
    ]
```

File: parser/testdata/secrets.json
Comment (spolti) [inline parser/testdata/secrets.json:1]:
```suggestion
```

File: parser/testdata/workflows/applicationrequest-issue69.json
Comment (spolti) [inline parser/testdata/workflows/applicationrequest-issue69.json:8]: can we have one without _file://_ to make sure it will work as before?

Comment (spolti) [reply parser/testdata/workflows/applicationrequest-issue69.json:8]: I mean, one test using the file without the protocol, as it was before, to make sure it still being unmarshaled correctly.

File: model/event_state.go
Comment (spolti) [inline model/event_state.go:59]:
```suggestion
// ApplyDefault set the default values for Event States
```

File: model/foreach_state.go
Comment (spolti) [inline model/foreach_state.go:92]:
```suggestion
// ApplyDefault set the default values for ForEach States
```

File: model/operation_state.go
Comment (spolti) [inline model/operation_state.go:55]:
```suggestion
// ApplyDefault set the default values for Operation States
```

File: model/parallel_state.go
Comment (spolti) [inline model/parallel_state.go:73]:
```suggestion
// ApplyDefault set the default values for Parallel States
```

File: model/states.go
Comment (spolti) [inline model/states.go:95]: will this be unmarshaled as part of the states?

Comment (ribeiromiranda) [reply model/states.go:95]: Yes. json.Unmarshal does the work.

**Review by spolti — COMMENTED** [review]
Just a few comments, otherwise looks good.
@lsytj0413 hi, do you want to take a final look before merging?

**Comment by spolti** [comment]
Thanks @lsytj0413!
@ribeiromiranda can you please address the open comments, including the ones from the sonatype-lift?

**Comment by ribeiromiranda** [comment]
@spolti I changed the line tag `// #nosec` to sonatype-lift ignore this check but it doesn't worked.

**Comment by spolti** [comment]
@ribeiromiranda both security issues are addressed in the go1.19.6.
The GH actions uses the latest 1.19.x release, I am not sure how to address this lift concerns.
We might ignore them for now.

File: model/util.go
Comment (spolti) [reply model/util.go:139]: https://github.com/serverlessworkflow/sdk-go/pull/159

Comment (spolti) [reply model/util.go:139]: @sonatype-lift ignore

Comment (spolti) [reply model/util.go:139]: @sonatype-lift ignore

**Comment by spolti** [comment]
> @spolti I changed the line tag `// #nosec` to sonatype-lift ignore this check but it doesn't worked.

File: model/util.go
Comment (spolti) [inline model/util.go:138]:
```suggestion
```

Comment (spolti) [reply model/util.go:138]: let's just remove it and move it forward.

**Review by spolti — APPROVED** [review]

File: model/util.go
Comment (ricardozanini) [inline model/util.go:117]: Can we have an issue for this one?

File: model/util.go
Comment (ricardozanini) [inline model/util.go:259]: Won't this be a problem? If not, we can remove this comment since it can be confusing for later maintainers.

File: parser/testdata/workflows/greetings-constants-file.sw.yaml
Comment (ricardozanini) [inline parser/testdata/workflows/greetings-constants-file.sw.yaml:22]: It should work without the schema (since the default is `file://`). Do we have a test to verify it?

File: model/util.go
Comment (ricardozanini) [inline model/util.go:108]: If no schema, then `file`? And then we can parse. So we won't require users to add the schema to their paths.

File: model/util.go
Comment (ricardozanini) [inline model/util.go:138]: In 0.9 we have the `auth` property that defines authentication parameters for this HTTP invocation. It's something we should do in the future. Just a note, not to implement at the moment.

**Review by ricardozanini — COMMENTED** [review]
Looks great! Just a few minor comments. I'm positive we will merge this soon.

Sorry for the late review, I was on PTO.

File: parser/testdata/workflows/greetings-constants-file.sw.yaml
Comment (ribeiromiranda) [reply parser/testdata/workflows/greetings-constants-file.sw.yaml:22]: Yes. Have the tests: `file://path/`, `file://../path/`, `./path/`, `/path/`

**Comment by ribeiromiranda** [comment]
@ricardozanini Changed the `getBytesFromFile` to `loadExternalResource` adding more validation.

File: model/util.go
Comment (ricardozanini) [reply model/util.go:177]: @sonatype-lift ignore

File: model/util.go
Comment (ricardozanini) [reply model/util.go:170]: @sonatype-lift ignore

File: model/util.go
Comment (ricardozanini) [reply model/util.go:177]: @sonatype-lift ignore

Comment (ricardozanini) [reply model/util.go:177]: @sonatype-lift ignore

**Review by ricardozanini — APPROVED** [review]

**Pull request merged**
